### PR TITLE
Completion with histchars and initial tilde

### PR DESCRIPTION
**Pull request by pghvlaans**
This PR addresses ~~three~~ four completion-related problems.

First: when using completion, any ~~exclamation marks~~ instance of the first `histchar` ("!" by default) will be left unescaped if it is included in the result, regardless of whether it had been escaped when the completion was performed:

    cd \!(tab)
    cd !very\ long\ and\ strange\ directory\ name/

This is perfectly correct if `histexpand` is turned off. If `histexpand` is enabled, however, the resulting command will either error out or perform an unexpected history expansion. To prevent this, escape any ~~exclamation marks~~ instance of the first `histchar` in the result of a completion if `histexpand` is enabled.

Second: if the third `histchar` ("#" by default) is anything other than `#`, it will not be escaped at the beginning of the string. This causes similar unexpected behavior.

~~Third: if a filename starts with a tilde (`~`), completion invariably fails unless it is escaped. Achieving a partial completion unescapes any tildes, making it necessary to navigate back and fix it. This happens because `addstar` is always set to zero when the completion string starts with a tilde; setting `addstar` to zero only when the tilde is by itself (or is entered as a directory name) solves the issue.~~ 

Third: if a filename starts with an escaped tilde (`~`), achieving a partial completion unescapes the first tilde, making it necessary to navigate back and fix it. This happens because `fmtx` only reaches `stakwrite` soon enough if an `S_EPAT` character is in the string, a category that includes `#` but not `~`. Alter the relevant loop to stop at `~` as well.

Fourth: when at the start of a quotation, `~` can complete to `$HOME` or undergo tilde expansion to an existing user. Send such `~` to `fmtx` so it can be escaped properly.

~~`completion.c`: In the function `fmtx`, add a preceding backslash to exclamation marks in the completion result if `histexpand` is set to `on`.~~

* `completion.c`: Escape initial tildes in completion candidates as intended, including tildes starting a quotation. In the function `fmtx`, when `histexpand` is on, add a preceding backslash to the first `histchar` everywhere, and to the third `histchar` when it appears at the beginning of the string.
* `hexpand.{c,h}`: Use a separate function to set the three histchars.

As a side note, `bash` escapes ~~exclamation marks~~ the first `histchar` in all cases, whereas `zsh` only escapes ~~exclamation marks~~ with `histexpand` enabled.

### CODE REVIEW

**Comment by McDutchie** [comment]
The history expansion characters can be changed using the `histchars` variable. Your patch does not seem to take this into account.

**Comment by pghvlaans** [comment]
It doesn't; I'll make some adjustments. Thank you!

**Comment by pghvlaans** [comment]
OK, that should do the trick. I'll update this comment when the regression tests are done.

EDIT: Done.

**Comment by pghvlaans** [comment]
Sorry; that was just rooting out a corner case where `#` and `~` are `histchar` 1 and 3, and 3 goes first.

**Comment by pghvlaans** [comment]
The `#`-related logic in `fmtx` is easier to follow and filename completion starting with `~` is fixed (see [top comment](https://github.com/ksh93/ksh/pull/639#issue-1689457846), third point).

**Comment by pghvlaans** [comment]
Well, that was more post-facto commits than I expected. Please let me know if I should close this PR and make a clean one.

**Comment by McDutchie** [comment]
Op 03-05-2023 om 04:58 schreef pghvlaans:
> Please let me know if I should close this PR and make a clean one.

No need, they'll all get squashed into one commit.

**Comment by pghvlaans** [comment]
OK, sounds good. I just didn't know whether having lots of commits made the process less convenient on your end.

**Comment by McDutchie** [comment]
I added a commit that streamlines things a bit. How does that look to you, @pghvlaans?

**Comment by pghvlaans** [comment]
I like it; `fmtx` in particular is much cleaner that way.

**Comment by McDutchie** [comment]
> I just didn't know whether having lots of commits made the process less convenient on your end.

It doesn't; I simply `git diff` the PR branch against the dev branch (or whatever commit it was branched off from) to review the differences. If anything I prefer to have the PR commits to refer back to as they can make individual changes easier to understand.

**Comment by McDutchie** [comment]
The fix for completion of file names beginning with `~` actually doesn't seem to work for me.

Steps:

1. `touch \~martijn`
2. `ls \~mar`[TAB]

The command line still reads `ls ~martijn` whereas it should read `ls \~martijn`.

I reverted my own changes to make sure I didn't break it, but that made no difference.

**Comment by pghvlaans** [comment]
That's working as intended, actually. `~` remains escaped if it's the only character, a directory name, or a histchar.

EDIT: The difference with `dev` in this specific case is that `ls ~mar`[TAB] completes successfully.

**Comment by McDutchie** [comment]
OK. I think that's a bug; I typed the backslash so I would expect it to stay there and not be turned into a tilde expansion. But we can treat that as a separate issue.

**Comment by pghvlaans** [comment]
My thought was that initial `~` doesn't turn into `$HOME` if there's something other than `/` following it, so it wouldn't need escaping in that situation (unless someone decided to turn it into histchar 0 or 2 for... whatever reason).

However, if that's undesirable, the following at line 53 would escape initial `~` that make it to `fmtx` without double-escaping `~` as a histchar:

    if((!hexp || (*cp!=hc[0] && *cp!=hc[2])) && (*cp=='#' || (*cp=='~' && *cp!=hc[0] && *cp!=hc[2])))

I'd just need to alter one of the regression tests.

**Comment by McDutchie** [comment]
But that change also causes a tilde to be escaped if I didn't type a backslash before it, which is indeed not desirable because tilde expansion is intended then.

I'll open a separate issue for this at some point.

**Comment by pghvlaans** [comment]
Whoops, I misunderstood. Yeah, it's definitely a bug.

**Comment by pghvlaans** [comment]
Completing something like `~martijn` without an escape definitely isn't good, so line 53 should be changed to:

    if((!hexp || (*cp!=hc[0] && *cp!=hc[2])) && (*cp=='#' || *cp=='~'))

That way, *if* such a filename were completed, it would be escaped (similar to behavior with filename-initial `#`).

The other question is whether filenames with unescaped tildes at the beginning should be allowed to complete at all. Removing `begin==last-1` from line 363 would stop that from happening.

**Comment by McDutchie** [comment]
[…]
> That way, _if_ such a filename were completed, it would be escaped (similar to behavior with filename-initial `#`).

Yes, this makes sense.

> The other question is whether filenames with unescaped tildes at the beginning should be allowed to complete at all.

In my view, the answer to that question is no. An unescaped tilde indicates an intention to perform tilde expansion. Some shells (bash, yash, zsh) support tilde expansion completion in that case. ksh doesn't (yet) but it still should not do the wrong completion.

**Comment by pghvlaans** [comment]
On reflection, agreed. Unexpected non-tilde expansion can be almost as dangerous as unexpected tilde expansion. I'll make the changes and push a commit.

**Comment by McDutchie** [comment]
Give me a moment, I'm working on something. I'll post a patch in a comment to avoid interfering.

**Comment by pghvlaans** [comment]
Ah, sorry. I saw your message just after pushing.

**Comment by McDutchie** [comment]
No problem, I wasn't really getting anywhere and I need to sleep now.

Maybe you can figure this out. Tilde expansion is happening in quoted completion contexts where it should do file name completion instead, regardless of `~`. We didn't introduce this, the problem exists on 93u+ 2012-08-01 as well.

Assuming user 'martijn' exists:

Type: `ls '~martijn`[TAB]    
Result: `$ ls /Users/martijn`

Same for `ls "~martijn`[TAB] and `ls $'~martijn`[TAB].

The culprit is the sh_argbuild() call on completion.c:391 which causes tilde expansion via this call stack (obtained by making tilde_expand2() abort):

```
  * frame #0: 0x00000001946a2d98 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x00000001946d7ee0 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000194612340 libsystem_c.dylib`abort + 168
    frame #3: 0x000000010006345c ksh`tilde_expand2(offset=17) at macro.c:2677:1
    frame #4: 0x000000010005b540 ksh`copyto(mp=0x00006000017002c0, endch=0, newquote=0) at macro.c:666:5
    frame #5: 0x000000010005c6f4 ksh`sh_macexpand(argp=0x000000010080f0e0, arghead=0x000000016fdfc6a0, flag=0) at macro.c:235:2
    frame #6: 0x000000010002bb5c ksh`arg_expand(argp=0x000000010080f0e0, argchain=0x000000016fdfc6a0, flag=0) at args.c:818:11
    frame #7: 0x000000010002b810 ksh`sh_argbuild(nargs=0x000000016fdfc798, comptr=0x000000010080f0a0, flag=0) at args.c:658:9
    frame #8: 0x0000000100032d18 ksh`ed_expand(ep=0x000000010080a600, outbuff="ls '~a", cur=0x000000010080a6fc, eol=0x000000010080a700, mode=92, count=-1) at completion.c:391:10
    frame #9: 0x000000010009aa14 ksh`escape(ep=0x00006000017040c0, out=L"\U0000fffd慾 '~a", count=-1) at emacs.c:974:7
    frame #10: 0x0000000100098aac ksh`ed_emacsread(context=0x000000010080a600, fd=0, buff="ls '~a", scend=1020, reedit=0) at emacs.c:625:13
```

**Comment by McDutchie** [comment]
Regression test patch that should pass after a fix:
```diff
diff --git a/src/cmd/ksh93/tests/pty.sh b/src/cmd/ksh93/tests/pty.sh
index edf54e50f..d829e815a 100755
--- a/src/cmd/ksh93/tests/pty.sh
+++ b/src/cmd/ksh93/tests/pty.sh
@@ -1219,6 +1219,15 @@ w cd chrtest3
 p :test-2:
 w ls \\~\t
 r ^:test-2: ls \\~ab/\r\n$
+p :test-3:
+w ls '~\t
+r ^:test-3: ls \\~ab/\r\n$
+p :test-4:
+w ls "~\t
+r ^:test-4: ls \\~ab/\r\n$
+p :test-5:
+w ls $'~\t
+r ^:test-5: ls \\~ab/\r\n$
 !
 
 # ======
```

**Comment by pghvlaans** [comment]
Looks like all we have to do there is go to `fmtx` if initial `~` is preceded by `'` or `"`. Line 363 to:

    if(*begin=='~' && !strchr(begin,'/') && begin[-1]!='\'' && begin[-1]!='"')
           addstar = 0;

The regression tests pass, but I wonder if the underlying bug was that it was supposed to be doing that anyway.

**Comment by McDutchie** [comment]
> Looks like all we have to do there is go to `fmtx` if initial `~` is preceded by `'` or `"`.

Nice find. That does appear to fix it. But I don't think it's quite enough; if you define a `.sh.tilde.get()` discipline function (see 936a1939a8ddb1aea60317ccf687e2cc4d520229) that outputs something to the terminal, you can see the tilde expansion is still executed, even though its result is discarded. Since we can now have tilde expansion discipline functions, tilde expansions may have side effects that should be avoided here. ~~I think I know how to fix that, though; in macro.c:666, we just need to stop `copyto` from calling `tilde_expand2` during file name completion, which means avoiding setting the `tilde` variable on line 445 if `sh_isstate(SH_FCOMPLETE)`. Commit coming up.~~

**Comment by McDutchie** [comment]
I think this is probably ready now.

**Comment by McDutchie** [comment]
Well, that commit was wrong: it causes regressions.
```
	pty.sh[450]: FAIL: vi mode file name completion: line 459: expected "^/tmp/fakehome_54117/testfile_54117\r?\n$", got EOF
	pty.sh[728]: FAIL: value of $? after tilde expansion in tab completion: line 737: expected "false /tmp", got EOF
```

**Comment by McDutchie** [comment]
Can you test the latest commit?

**Comment by pghvlaans** [comment]
No fails on Linux. I'll go ahead and do my usual testing round.

EDIT: Got two delay-related fails on macOS.

**Comment by pghvlaans** [comment]
Other than the delay increase I mentioned earlier, everything looks fine.

**Comment by McDutchie** [comment]
Brilliant, thanks for the PR!

**Comment by pghvlaans** [comment]
No worries!
